### ISSUE TITLE
Obfs4 optimization

### ISSUE THREAD

**Issue by keyboardcat1** [guardianproject/orbot-apple#71]
I am currently in China and have to use obfs4 bridges to connect to tor. However, I rarely can manage, sometimes having to wait a day or even two of repeatedly waiting 5-10 minutes, clearing cache and switching (previously proved functional) bridges. The progress stays stuck mostly on 50% and tor logs get flooded with "general SOCKS server failure". I believe there is a lot of room for improvement since the app "Tor Browser" manages to connect to the same bridges in under a minute.

**Comment by tladesignz** [comment]
Uuuuh. Hello! Good to talk to you!

This is very interesting information, however it would be more helpful, if you could provide some more detail.

> since the app "Tor Browser" manages to connect to the same bridges in under a minute.

Can you point me to the location of where you got the "Tor Browser" app? Are you talking about the desktop computer app provided by Tor Project for Mac, Windows and Linux? Or is this an iOS app officially available on the Apple App Store for iOS?

The official recommendation for mainland China is to use Snowflake instead of Obfs4 proxies. It is well known, that the "great firewall" blocks most Obfs4 proxies pretty quickly.

Can you cross-check with your "Tor Browser" app, that you're actually using Obfs4 proxies, and not Snowflake?

Can you use Orbot iOS with Snowflake and tell me, what's the performance with this? Please try both options (traditional and AMP)!

Thank you!

**Comment by keyboardcat1** [comment]
No, Tor Browser is an app unrelated to Orbot. I am bringing i up because of how faster it somehow is.

**Comment by n8fr8** [comment]
My point is that it may not be Tor at all. There are many fakes, imposters, and closed source scam apps out there claiming to be Tor. Can you provide a link for us to look at so we can investigate?

**Comment by n8fr8** [comment]
What this app may be doing is a traditional VPN to their servers, and THEN routing that traffic over Tor. That would make it faster, but obviously not true Tor to the client on your device.

The really is no way to know since it is closed source, their website is down, and their contact info likely doesn't work.

**Comment by keyboardcat1** [comment]
Bridges are required here in China to connect through the app. They could be faking that need for bridges through geolocation, but I highly doubt that is the case since they have both tor and tor + vpn options

**Comment by keyboardcat1** [comment]
Thanks a lot for the explanations! I'm excited for Tor in Rust, on mobile on top of that.

I'm a still bit confused though, how does it manage to connect faster to bridges than Orbot? What corners does it cut?

**Comment by tladesignz** [comment]
> I'm a still bit confused though, how does it manage to connect faster to bridges than Orbot? What corners does it cut? Is it because it isn't a network extension (doesn't apply globally) and hence isn't restricted in terms of memory?

Yeah. We need to throttle Tor's memory usage. There's unfortunately not a lot of ways we can do that. In fact, there's only one option: [MaxMemInQueues](https://2019.www.torproject.org/docs/tor-manual.html.en#MaxMemInQueues)

I have a feeling from the reports I get, that this poses a problem in censored areas. Maybe, the constraint environment makes Tor run out of memory there, because it needs to work more for getting a connection.

Maybe we need to make this configurable. 🤔

**Comment by tladesignz** [comment]
> Maybe we need to make this configurable. 

#72

**Comment by keyboardcat1** [comment]
Also, is it normal behavior that there is a ton of general SOCKS errors when using obsf4? As I understand it, it is the bridge failing to connect to relays, as opposed to the client being unable to connect to the bridge, which would imply that the client is attempting to connect to invalid tor relays.
I retried original snowflake, and for some reason it connects very reliably and quickly now (I think I forgot to clear cache last time). However, the connection is comparably slower to obfs4 despite the massive number of messages. Also, the WebSocket connection gets cut every 20s or so but successfully reconnects to the same bridge.

**Comment by tladesignz** [comment]
> Also, is it normal behavior that there is a ton of general SOCKS errors when using obsf4? 

No, that's not normal, but I hear reports from others describing this. So probably due to the censorship.

> As I understand it, it is the bridge failing to connect to relays, as opposed to the client being unable to connect to the bridge, which would imply that the client is attempting to connect to invalid tor relays.

That doesn't sound right. Bridges run their own entry node and they're typically placed in a location which is connected well.
Anyhow, I'm really no expert there, sorry.

> I retried original snowflake, and for some reason it connects very reliably and quickly now (I think I forgot to clear cache last time). However, the connection is comparably slower to obfs4 despite the massive number of messages. Also, the WebSocket connection gets cut every 20s or so but successfully reconnects to the same bridge.

This is very interesting info, thanks!

**Comment by keyboardcat1** [comment]
GitHub apparently is unblocked now (or was never blocked but I assumed it was) so I'll post this while I can.

Unless the GUI or the rest of Tor is very memory intensive, it looks like MaxMemInQueues has been misconfigured. With a restriction of 50 MB, one would expect MaxMemInQueues to be in that order of magnitude. However, it seems like it has been set to around 5.2 MB (or 5 MiB).
Here's what I'm talking about:
```txt
{GENERAL} We're low on memory (cell queues total alloc: 528 buffer total alloc: 5251072, tor compress total alloc: 0 (zlib: 0, zstd: 0, lzma: 0), rendezvous cache total alloc: 0). Killing circuits withover-long queues. (This behavior is controlled by MaxMemInQueues.)
Jul 31 05:07:52.000 [notice] {GENERAL} Removed 5218832 bytes by killing 1 circuits; 0 circuits remain alive. Also killed 0 non-linked directory connections. Killed 0 edge connections
```

Circuits get killed as soon as the sum of their allocated memory exceeds around 5 MB. This would explain the pattern I observe whenever I use bridges: Tor builds up microdescriptors, exceeds MaxMemInQueues, gets reset and so on, so that microdescriptors rarely exceed a count of 200, and stays stuck at 50%.

The real question should be how can it even connect in the first place? It must get extremely lucky to get the correct microdescriptors to build a circuit before it gets killed.

**Comment by tladesignz** [comment]
Ah, thank you for digging that up. Yeah, as I expected, it's the memory limitation.

The memory accounting is not that easy, though.

- UI doesn't play a role here, that's in the app alone, which has nothing to do with the so called "Network Extension", which is it's own process and runs in the background. (A rare case, where iOS apps can run more than one process. However, in a very limited way.)

- Queue memory isn't the only memory the Network Extension uses.
- We need to leave headroom for other workings of Tor, which we cannot memory limit directly.
- We need to leave headroom for Obfs4proxy or Snowflake.

I played around with #72 again, and it remembered me, why I set it to 5 MB in the first place:

As soon as I set it higher, memory usage explodes during startup and exceeds 50 Mbyte, which makes it getting shot off by Apple's Jetsam memory accounting daemon immediately again.

I will, however, push changes which will let this setting be overridable by advanced users anyway, so users in difficult environments (like you) will be able to play with it and maybe discover settings which works for them. In the hope they'll report their findings.

I'll talk to Tor engineers, too. Maybe they can find a way to throttle memory consumption during startup, so we can improve the situation until Arti (A Rust Tor Implementation) arrives.

**Comment by tladesignz** [comment]
You can now override `MaxMemInQueues`.
Fixes to the startup memory consumption are on their way to getting published, too.
